internal/internal_task_handlers.go

```diff
@@ -262,6 +262,28 @@ func isDecisionEvent(eventType s.EventType) bool {
 	}
 }
 
+// isDecisionEventForReplay is different from isDecisionEvent because during replays
+// we want to intentionally ignore workflow complete/fail/cancel/continueasnew events so that
+// decision tree replays matches with the workflow processing respond tasks
+func isDecisionEventForReplay(eventType s.EventType) bool {
+	switch eventType {
+	case
+		s.EventTypeActivityTaskScheduled,
+		s.EventTypeActivityTaskCancelRequested,
+		s.EventTypeTimerStarted,
+		s.EventTypeTimerCanceled,
+		s.EventTypeCancelTimerFailed,
+		s.EventTypeMarkerRecorded,
+		s.EventTypeStartChildWorkflowExecutionInitiated,
+		s.EventTypeRequestCancelExternalWorkflowExecutionInitiated,
+		s.EventTypeSignalExternalWorkflowExecutionInitiated,
+		s.EventTypeUpsertWorkflowSearchAttributes:
+		return true
+	default:
+		return false
+	}
+}
```

Comment:
To stick a comment in here from chats a while back, for visibility:
I _generally like_ this approach and I think it stands a pretty good chance of being an improvement...

... but I'm not (yet) confident that it's _correct_, nor that it doesn't worsen other scenarios.  It's a pretty substantial change despite how small it is.  From vague memory of my last attempt to verify, I ended up with more concerns than I started with, but no real evidence in either direction :\

It may be fine even if it does introduce new flaws, it's definitely not perfect now, but we need to have an informed decision.  And we now have a LOT more internal replay-shadow-tests (compared to when this was first made) that we can run to discover what has changed - if it finds nothing but improvements against our fairly wide range of behavior in those tests, that will probably be good enough.
I just _definitely_ do not trust our test suite here to _prove correctness_, they don't even use like half of our features nor do they use them in complicated ways.  Issues they find are plenty useful, but they're very strange in a lot of ways and that makes them very very far from evidence of correctness.

---

So tl;dr: it just has to be sufficiently-validated or fully understood/described.
We've done some validation (thanks @agautam478 !) but we need to do a bit more, and none of the tests here seem like particularly strong evidence for _what changed_ so I need to dig back in and try to reason about it more carefully.

Reply:
All sounds good to me. At the time I looked into these false positives I focused on a subset of the potential scenarios. I'm not super confident this is completely safe change even though it looks like it is. Finding more scenarios and adding them as test cases should be next step as discussed offline. Due to other priorities I'm handing this over to you and @agautam478.